pymc3/tests/test_variational_inference.py

```diff
@@ -0,0 +1,123 @@
+import unittest
+import numpy as np
+import theano
+import theano.tensor as tt
+from theano.sandbox.rng_mrg import MRG_RandomStreams
+import pymc3 as pm
+from pymc3 import Model, Normal
+from pymc3.variational.opvi import KL, MeanField, FullRank, TestFunction
+from ..theanof import set_tt_rng
+from . import models
+set_tt_rng(MRG_RandomStreams(42))
+
+
+class TestApproximates:
+    class Base(unittest.TestCase):
+        op = KL
+        approx = MeanField
+        test_function = TestFunction
+        NITER = 30000
+
+        def test_elbo(self):
+            if self.approx is not MeanField:
+                raise unittest.SkipTest
+            mu0 = 1.5
+            sigma = 1.0
+            y_obs = np.array([1.6, 1.4])
+
+            post_mu = np.array([1.88])
+            post_sd = np.array([1])
+            # Create a model for test
+            with Model() as model:
+                mu = Normal('mu', mu=mu0, sd=sigma)
+                Normal('y', mu=mu, sd=1, observed=y_obs)
+
+            # Create variational gradient tensor
+            mean_field = self.approx(model=model)
+            elbo = -self.op(mean_field)(self.test_function())(mean_field.random())
+
+            mean_field.shared_params['mu'].set_value(post_mu)
+            mean_field.shared_params['rho'].set_value(np.log(np.exp(post_sd) - 1))
+
+            f = theano.function([], elbo)
+            elbo_mc = sum(f() for _ in range(10000))/10000
+
+            # Exact value
+            elbo_true = (-0.5 * (
+                3 + 3 * post_mu**2 - 2 * (y_obs[0] + y_obs[1] + mu0) * post_mu +
+                y_obs[0]**2 + y_obs[1]**2 + mu0**2 + 3 * np.log(2 * np.pi)) +
+                0.5 * (np.log(2 * np.pi) + 1))
+            np.testing.assert_allclose(elbo_mc, elbo_true, rtol=0, atol=1e-1)
+
+        @unittest.skip('Not ready')
+        def test_vary_samples(self):
+            _, model, _ = models.simple_model()
+            i = tt.iscalar('i')
+            i.tag.test_value = 1
+            with model:
+                mf = self.approx()
+                elbo = mf.elbo(i)
+            elbos = theano.function([i], elbo)
+            self.assertEqual(elbos(1).shape[0], 1)
+            self.assertEqual(elbos(10).shape[0], 10)
```

Comment:
I've copied this file from #1600, interface is different so tests will fail. I'll refactor them later when interface will be convenient